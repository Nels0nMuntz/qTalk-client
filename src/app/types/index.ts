export * from './createSubtalkResponse.interface'
export * from './subscriptionResponse.interface'