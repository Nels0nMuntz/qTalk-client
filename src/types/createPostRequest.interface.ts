import { CreatePostPayload } from '@/lib/validators/postSchema';

export interface CreatePostRequest extends CreatePostPayload {}
