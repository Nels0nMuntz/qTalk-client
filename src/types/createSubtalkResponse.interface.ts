export interface CreateSubtalkResponse {
  name: string;
}
