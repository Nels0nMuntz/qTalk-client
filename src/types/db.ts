import { Comment, Post, Subtalk, User, Vote } from '@prisma/client';

export type ExtendedPost = Post & {
  subtalk: Subtalk;
  votes: Vote[];
  author: User;
  comments: Comment[];
};
