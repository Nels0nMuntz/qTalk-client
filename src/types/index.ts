export * from './createSubtalkResponse.interface';
export * from './subscriptionResponse.interface';
export * from './createPostRequest.interface';
export * from './searchResult.interface';
export * from './db';
export * from './redis';
