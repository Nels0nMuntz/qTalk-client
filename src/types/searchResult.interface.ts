import { Prisma, Subtalk } from '@prisma/client';

interface ExtendedSubtalk extends Subtalk {
  _count: Prisma.SubtalkCountOutputType;
}

export interface SearchResult extends Array<ExtendedSubtalk> {}
