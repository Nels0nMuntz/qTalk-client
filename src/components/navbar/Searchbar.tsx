'use client';

import React, { useCallback, useEffect, useRef, useState } from 'react';
import { useQuery } from '@tanstack/react-query';
import axios from 'axios';
import { usePathname, useRouter } from 'next/navigation';
import debounce from 'lodash.debounce';
import {
  Command,
  CommandEmpty,
  CommandGroup,
  CommandInput,
  CommandItem,
  CommandList,
} from '../ui/command';
import { SearchResult } from '@/types';
import { Users } from 'lucide-react';

export default function Searchbar() {
  const router = useRouter();
  const pathname = usePathname();
  const commandRef = useRef<HTMLDivElement>(null);
  const [value, setValue] = useState<string>('');
  const {
    data: queryResult,
    refetch,
    isFetched,
  } = useQuery({
    queryFn: async () => {
      if (!value) return [];

      const { data } = await axios.get<SearchResult>(`/api/search?q=${value}`);
      return data;
    },
    queryKey: ['search'],
    enabled: false,
  });

  useEffect(() => {
    setValue('');
  }, [pathname]);

  const request = debounce(async () => {
    refetch();
  }, 400);

  const debounceRequest = useCallback(() => {
    request();
  }, []);

  return (
    <Command
      ref={commandRef}
      className="relative rounded-lg border max-w-lg z-50 overflow-visible"
    >
      <CommandInput
        onValueChange={(text) => {
          setValue(text);
          debounceRequest();
        }}
        value={value}
        className="outline-none border-none focus:border-none focus:outline-none ring-0"
        placeholder="Search communities..."
      />
      {Boolean(value.length) && (
        <CommandList className="absolute bg-white top-full inset-x-0 shadow rounded-b-md">
          {isFetched && <CommandEmpty>No results found.</CommandEmpty>}
          {Boolean(queryResult?.length) && (
            <CommandGroup heading="Communities">
              {queryResult?.map(({ id, name }) => (
                <CommandItem
                  key={id}
                  value={name}
                  onSelect={(value) => {
                    router.push(`/t/${value}`);
                    router.refresh();
                  }}
                >
                  <Users className="w-4 h-4 mr-2" />
                  <a href={`/t/${name}`}>t/{name}</a>
                </CommandItem>
              ))}
            </CommandGroup>
          )}
        </CommandList>
      )}
    </Command>
  );
}
