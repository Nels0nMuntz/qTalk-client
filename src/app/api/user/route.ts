import { db } from '@/lib/db';
import { buildResponse } from '@/lib/utils';
import { signUpFormSchema } from '@/lib/validators';
import { hash } from 'bcrypt';
import { z } from 'zod';

export async function POST(req: Request) {
  try {
    const body = await req.json();
    const { email, username, password } = signUpFormSchema.parse(body);

    const userByEmail = await db.user.findFirst({
      where: {
        email,
      },
    });
    const userByUsername = await db.user.findFirst({
      where: {
        username,
      },
    });
    if (userByEmail || userByUsername) {
      return buildResponse('Email or username are taken', { status: 422 });
    }

    const hashedPassword = await hash(password, 10);
    await db.user.create({
      data: {
        email,
        username,
        password: hashedPassword,
      },
    });
    return buildResponse('User was created', { status: 201 });
  } catch (error) {
    console.log(error);

    if (error instanceof z.ZodError) {
      return new Response(error.message, { status: 422 });
    }

    return buildResponse('Internal server error', { status: 500 });
  }
}
