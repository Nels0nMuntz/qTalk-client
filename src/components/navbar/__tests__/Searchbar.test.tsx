import { render, screen } from "@testing-library/react";
import user from "@testing-library/user-event";
import { useQuery } from "@tanstack/react-query";
import Searchbar from "../Searchbar";
import { SearchResult } from "@/types";

// const pushMock = jest.fn();
// const refreshMock = jest.fn();
jest.mock('next/navigation', () => ({
  useRouter: jest.fn(() => ({
    push: jest.fn(),
    refresh: jest.fn(),
  })),
  usePathname: jest.fn(),
}));

jest.mock('@tanstack/react-query', () => ({
  useQuery: jest.fn(() => ({
    data: {},
    refetch: jest.fn(),
    isFetched: false,
  })), 
}));

jest.mock('axios', () => ({}));


describe('Searchbar component', () => {
  it('does not render listbox until type any value', async () => {
    const { container } = render(<Searchbar/>);

    const input = await screen.findByRole('combobox');
    const listbox = screen.queryByRole('listbox');

    expect(input).toHaveValue('');
    expect(listbox).not.toBeInTheDocument();
    // expect(container).toMatchSnapshot();
  });
  it('renders listbox after a value was provided', async () => {
    const getServerSessionMock = useQuery as jest.MockedFunction<typeof useQuery<SearchResult>>;
    getServerSessionMock.mockResolvedValueOnce();
    const { container } = render(<Searchbar/>);

    const input = await screen.findByRole('combobox');
    const listbox = await screen.findByRole('listbox');

    await user.type(input, 'react');

    expect(input).toHaveValue('react');
    expect(listbox).toBeInTheDocument();
    // expect(container).toMatchSnapshot();
  });
})