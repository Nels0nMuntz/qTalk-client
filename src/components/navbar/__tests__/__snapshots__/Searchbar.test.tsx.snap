// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Searchbar component does not render listbox until type any value 1`] = `
<div>
  <div
    class="flex h-full w-full flex-col bg-popover text-popover-foreground relative rounded-lg border max-w-lg z-50 overflow-visible"
    cmdk-root=""
  >
    <label
      cmdk-label=""
      for=":r2:"
      id=":r1:"
      style="position: absolute; width: 1px; height: 1px; padding: 0px; margin: -1px; overflow: hidden; clip: rect(0px, 0px, 0px, 0px); white-space: nowrap; border-width: 0px;"
    />
    <div
      class="flex items-center border-b px-3"
      cmdk-input-wrapper=""
    >
      <svg
        class="mr-2 h-4 w-4 shrink-0 opacity-50"
        fill="none"
        height="24"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <circle
          cx="11"
          cy="11"
          r="8"
        />
        <path
          d="m21 21-4.3-4.3"
        />
      </svg>
      <input
        aria-autocomplete="list"
        aria-controls=":r0:"
        aria-expanded="true"
        aria-labelledby=":r1:"
        autocomplete="off"
        autocorrect="off"
        class="flex h-11 w-full rounded-md bg-transparent py-3 text-sm placeholder:text-muted-foreground disabled:cursor-not-allowed disabled:opacity-50 outline-none border-none focus:border-none focus:outline-none ring-0"
        cmdk-input=""
        id=":r2:"
        placeholder="Search communities..."
        role="combobox"
        spellcheck="false"
        type="text"
        value=""
      />
    </div>
  </div>
</div>
`;
